File: dynamic_programming/recursive_optimal_binary_search_tree.py
import sys


def optimal_bst(p, q, idx_of_start, idx_of_end):
    if idx_of_end == idx_of_start - 1:
        return q[idx_of_start - 1]

    w = q[idx_of_start - 1]
    for l in range(idx_of_start, idx_of_end + 1):
        w += p[l]
        w += q[l]
    w = round(w, 2)

    result = sys.maxsize
    for r in range(idx_of_start, idx_of_end + 1):
        result = min(result, optimal_bst(p, q, idx_of_start, r - 1) \
            + optimal_bst(p, q, r + 1, idx_of_end) + w)

    return round(result, 2)


if __name__ == '__main__':
    p = [0.0, 0.15, 0.10, 0.05, 0.10, 0.20]
    q = [0.05, 0.10, 0.05, 0.05, 0.05, 0.10]

    print("  i |", end='')
    for i in range(len(p)):
        print(f"{i:^5}|", end='')
    print('')
    print("p[i]|", end='')
    for _p in p:
        print(f"{_p:>5.2f}|", end='')
    print('')
    print("q[i]|", end='')
    for _q in q:
        print(f"{_q:>5.2f}|", end='')
    print('')

    diff = -1
    while diff < len(p):
        print(f"--- diff = {diff} ---")

        i = 1
        j = i + diff
        while j < len(p):
            result = optimal_bst(p, q, i, j)

            print(f"i = {i}, j = {j} / result = {result:.2f}")

            i += 1
            j += 1

        diff += 1

File: dynamic_programming/recursive_optimal_binary_search_tree.go
//go:build ignore

package main

import (
	"fmt"
	"math"
)

func optimalBst(p, q []float32, i, j int) float32 {
	if j == i-1 {
		return q[i-1]
	}

	w := q[i-1]
	for l := i; l <= j; l += 1 {
		w += p[l]
		w += q[l]
	}

	result := float32(math.MaxInt)
	for r := i; r < j+1; r += 1 {
		result = min(result, optimalBst(p, q, i, r-1)+optimalBst(p, q, r+1, j)+w)
	}

	return result
}

func main() {
	p := []float32{0, 0.15, 0.10, 0.05, 0.10, 0.20}
	q := []float32{0.05, 0.10, 0.05, 0.05, 0.05, 0.10}

	fmt.Print("  i |")
	for i := 0; i < len(p); i += 1 {
		fmt.Printf("%4d|", i)
	}
	fmt.Println()
	fmt.Print("p[i]|")
	for i := 0; i < len(p); i += 1 {
		fmt.Printf("%.2f|", p[i])
	}
	fmt.Println()
	fmt.Print("q[i]|")
	for i := 0; i < len(q); i += 1 {
		fmt.Printf("%.2f|", q[i])
	}
	fmt.Println()

	for diff := -1; diff < len(p)-1; diff += 1 {
		fmt.Printf("--- diff = %d ---\n", diff)

		i := 1
		for j := i + diff; j < len(p); j += 1 {
			result := optimalBst(p, q, i, j)

			fmt.Printf("(i, j) = (%d, %d) / result = %.2f\n", i, j, result)

			i += 1
		}
	}
}
